product_variant_default_code/models/product.py

```
@@ -15,30 +15,91 @@
+
+
+def extract_token(s):
+    pattern = re.compile(r'\[([^\]]+?)\]')
+    return set(pattern.findall(s))
+
+
+def sanitize_reference_mask(product, mask):
+    tokens = extract_token(mask)
+    attribute_names = set()
+    for line in product.attribute_line_ids:
+        attribute_names.add(line.attribute_id.name)
+    if not tokens.issubset(attribute_names):
+        raise except_orm(_('Error'), _('Found unrecognized attribute name in '
+                                       '"Variant Reference Mask"'))
+
+
+def render_default_code(product, mask):
+    product_attrs = defaultdict(str)
+    reference_mask = ReferenceMask(mask)
+    for value in product.attribute_value_ids:
+        if value.attribute_code:
+            product_attrs[value.attribute_id.name] += value.attribute_code
+    all_attrs = extract_token(mask)
+    missing_attrs = all_attrs - set(product_attrs.keys())
+    missing = dict.fromkeys(missing_attrs, PLACE_HOLDER_4_MISSING_VALUE)
+    product_attrs.update(missing)
+    default_code = reference_mask.safe_substitute(product_attrs)
+    product.default_code = default_code
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    default_prefix = fields.Char(
-        string='Reference prefix',
-        help='Prefix for using when building internal references for the '
-             'variants generated from this template')
+    reference_mask = fields.Char(
+        string='Variant reference mask',
+        help='Reference mask for building internal references of a'
+             'variant generated from this template.'
+             'eg. "PREFIX-[r]-[d]-[f]-APPENDIX" where the "[]" surrounded '
+             'string like "r,d,f" are attribute name for the current product '
+             'When rendering, the "[xxx]" part will be replaced '
+             'by the corresponding code of attribute value for the variant.\n'
+             'Note: make sure characters "[,]" do not appear in your '
+             'attribute name')
```

We should provide a default value for this mask, because on contrary, there won't be differentiation on variant references. Doesn't it?